iot_service.go

```diff
@@ -1,181 +1,100 @@
-					logrus.WithField("err", err).Error("failed_to_write_message_to_client")
-				}
-			} else {
-				logrus.WithField("req", req).Error("instance_does_not_exist")
-			}
-		default:
-			// status switching
-			if instance, ok := chargerIdMap[req.ChargerId]; ok {
-				instance.status = req.Status
-				chargerIdMap[req.ChargerId] = instance
-				if err := c.WriteMessage(mt, []byte("instance status update success")); err != nil {
-					logrus.WithField("err", err).Error("failed_to_write_message_to_client")
-				}
-			} else {
-				err := c.WriteMessage(mt, []byte("instance not registered"))
-				logrus.WithField("err", err).Error("failed_to_write_message_to_client")
-			}
-
-		}
-		mapLock.Unlock()
-	}
-}
-
-func main() {
-	flag.Parse()
-	http.HandleFunc("/ws", wsEndpoint)
-	http.HandleFunc("/charger", getChargerStatus)
-	http.HandleFunc("/iot/health", getServiceHealthCheck)
-
-	logrus.Info("serving_iot_service...")
-	err := http.ListenAndServe(*addr, nil)
-	if err != nil {
-		log.Fatal("ListenAndServe: ", err)
-	}
-}
+package main
+
+import (
+	"flag"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/sirupsen/logrus"
+
+	"github.com/gin-contrib/cors"
+	"github.com/gin-gonic/gin"
+
+	"github.com/NUS-EVCHARGE/iot-service/handler"
+)
+
+var (
+	r *gin.Engine
+)
+
+var addr = flag.String("addr", ":8080", "http service address")
+var chargerIdMap = map[string]Instance{}
+var mapLock sync.Mutex
+
+var upgrader = websocket.Upgrader{
+	ReadBufferSize:  1024,
+	WriteBufferSize: 1024,
+	CheckOrigin:     func(r *http.Request) bool { return true },
+} // use default options
+
+type IoTHubRequest struct {
+	Command   string `json:"command"`
+	CompanyId string `json:"company_id"`
+	ChargerId string `json:"charger_id"`
+	Status    Status `json:"status"`
+}
+
+type Instance struct {
+	conn        *websocket.Conn
+	ChargerId   string
+	status      Status // available, charging
+	messageType int
+}
+
+type Status int
+
+const (
+	Available Status = iota
+	Pending
+	Charging
+	Offline
+	Error
+)
+
+func main() {
+
+	// self-signed cert and key in same folder to support https and wss
+	tlsCert := "mkcert.pem"
+	tlsKey := "mkkey.pem"
+
+	flag.Parse()
+
+	r = gin.Default()
+	r.Use(cors.New(cors.Config{
+		AllowOrigins:     []string{"*", "https://localhost:443"},
+		AllowMethods:     []string{"PUT", "PATCH", "POST", "GET", "DELETE", "OPTIONS", "HEAD"},
+		AllowHeaders:     []string{"authorization", "access-control-allow-origin", "Access-Control-Allow-Headers", "Origin", "Content-Length", "Content-Type", "authentication", "Access-Control-Request-Method", "Access-Control-Request-Headers"},
+		ExposeHeaders:    []string{"Content-Length"},
+		AllowCredentials: true,
+		MaxAge:           12 * time.Hour,
+	}))
+	registerHandler()
+
+	// amended r.Run to r.RunTLS with additional arguments to implement https and wss
+	if err := r.RunTLS(":443", tlsCert, tlsKey); err != nil {

```

Comment:
This is creating 443 even is running locally. Is it because you guys want to create secured web socket on AWS? A bit weird local hitting 443

@gitkh2891 , @sweiyang